File: 337.打家劫舍-iii.c
/*
 * @lc app=leetcode.cn id=337 lang=c
 *
 * [337] 打家劫舍 III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int *robRange(struct TreeNode *root)
{
  if (root == NULL)
  {
    int *dpn = (int *)malloc(sizeof(int) * 2);
    dpn[0] = 0;
    dpn[1] = 0;
    return dpn;
  }

  int *leftVal = robRange(root->left);
  int *rightVal = robRange(root->right);

  int val1 = root->val + leftVal[0] + rightVal[0];
  int maxl = leftVal[0] > leftVal[1] ? leftVal[0] : leftVal[1];
  int maxr = rightVal[0] > rightVal[1] ? rightVal[0] : rightVal[1];
  int val2 = maxl + maxr;
  int *dp = (int *)malloc(sizeof(int) * 2);
  dp[1] = val1;
  dp[0] = val2;

  return dp;
}

int rob(struct TreeNode *root)
{
  int *dp = robRange(root);
  return dp[0] > dp[1] ? dp[0] : dp[1];
}
// @lc code=end

File: store/337.打家劫舍-iii.js
/*
 * @lc app=leetcode.cn id=337 lang=javascript
 *
 * [337] 打家劫舍 III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
const traverse = (node) => {
  if (!node) return [0, 0];

  let left = traverse(node.left);
  let right = traverse(node.right);
  let noStolen = Math.max(left[1], left[0]) + Math.max(right[0], right[1]);
  let stolen = node.val + left[0] + right[0];
  return [noStolen, stolen];
};
var rob = function (root) {
  return Math.max(...traverse(root));
};
// @lc code=end
